text/0000-grand-testing-unification.md

```diff
@@ -0,0 +1,647 @@
+
+  await this.render(hbs`{{primitive-select items=listing}}`);
+
+  let options = this.find('option');
+  assert.equal(options.length, 4);
+  assert.equal(options.textContent.trim(), 'onetwo34');
+});
+
+test('provided selected item is selected', async function(assert) {
+  this.set('listing', Ember.A(['one', 'two', '3', '4']));
+  this.set('selectedItem', 'two');
+
+  await this.render(hbs`{{primitive-select selected=selectedItem items=listing}}`);
+
+  let [selectedOption, ...rest] = this.findAll('option:selected');
+  assert.equal(rest.length, 0, 'only one selected option found');
+
+  assert.equal(selectedOption.textContent.trim(), 'two', 'correct option is selected');
+});
+
+test('when changed fires update action with new value', async function(assert) {
+  assert.expect(1);
+
+  this.on('checkValue', function(newValue) {
+    assert.equal(newValue, 'two');
+  });
+
+  this.set('listing', Ember.A(['one', 'two', '3', '4']));
+
+  await this.render(hbs`{{primitive-select update=(action "checkValue") items=listing}}`);
+
+  await this.fillIn('select', 'two');
+});
+```
+
+# Pedagogy
+
+The main changes that this RFC makes to pedagogy is to massively simplify the set of testing special cases that consumers have to manage mentally while developing an application. The divide between acceptance tests and unit/integration tests largely vanishes, and we begin leveraging language features where we previously implemented a custom DSL.
+
+It is extremely important that as these changes land in the various libraries, we ensure the testing guides listed on guides.emberjs.com are kept up to date. A rewrite of large sections of the existing guide are likely needed, but this rewrite should result in a massive improvement due to the conceptual overhead that is being removed.
+
+# Drawbacks
+
+- This is a relatively large set of changes that are arguably not needed (things mostly work today).
+- To debug  `async` / `await` we would increasingly rely on sourcemap accuracy. This might negatively impact debuggability.
+- One of the major pains in upgrading larger applications to newer Ember versions, is updating their tests to follow "new" patterns.  This RFC introduces yet another "new" thing (and proposes to deprecate the old thing), and could therefore be considered "just more churn".
+
+# Alternatives
+
+* The primary alternative to this would be to do nothing...
+* Add unit/integration async test helpers without unifying their API with todays acceptance testing system, and avoid deprecating any existing API's
+
+# Unresolved questions
+
+- Should we prefer `getOwner(this)` to `this.owner`?  I believe that accessing the owner in testing is much more common (for mocking/stubbing, looking up singleton objects, etc), so we should use `this.owner`.
+- Should `this.find` return a jQuery wrapped element? I would prefer to stick with the main DOM API's here, so that we have a chance to share tests between the Fastboot and Browser.
```

Comment:
@rwjblue regarding…

> Should this.find return a jQuery wrapped element? I would prefer to stick with the main DOM API's here, so that we have a chance to share tests between the Fastboot and Browser.

We worked on https://github.com/cibernox/ember-native-dom-helpers/releases/tag/v0.2.0 cc\ @cibernox and decided to only return `HTMLElement` or `NodeList` from `find` helper. If a test must use jQuery the developer can use `jQuery(find('a'))` explicitly.